### PR TITLE
deal.II package: Conflicts between SymEngine and Trilinos build modes

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -157,6 +157,13 @@ class Dealii(CMakePackage, CudaPackage):
     depends_on('trilinos@master+amesos+aztec+epetra+ifpack+ml+muelu+rol+sacado+teuchos~amesos2~ifpack2~intrepid2~kokkos~tpetra~zoltan2',       when='+trilinos+mpi~int64+cuda')
     depends_on('trilinos@master+amesos+aztec+epetra+ifpack+ml+muelu+rol+sacado+teuchos~hypre~amesos2~ifpack2~intrepid2~kokkos~tpetra~zoltan2', when='+trilinos+mpi+int64+cuda')
 
+    # Both Trilinos and SymEngine bundle the Teuchos RCP library.
+    # This leads to conflicts between macros defined in the included
+    # headers when they are not compiled in the same mode.
+    # See https://github.com/symengine/symengine/issues/1516
+    depends_on("symengine~debug", when="^symengine ^trilinos~debug")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: do you need `^symengine` here? I think `when="^trilinos~debug")` should be enough.

File: var/spack/repos/builtin/packages/symengine/package.py [hunk 2]
```diff
@@ -38,9 +40,6 @@ class Symengine(CMakePackage):
             description='Enable thread safety option')
     variant('shared',       default=True,
             description='Enables the build of shared libraries')
-    variant('build_type', default='Release',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/symengine/package.py]: `build_type` is a standard `variant` for `CmakePackage`, here it's overwritten to only support two options. I think `build_type=Release` is the right way. 

I am not sure, but I hope 
```
depends_on('symengine build_type=debug',...)
```
works in `dealii`.

File: var/spack/repos/builtin/packages/symengine/package.py [hunk 3]
```diff
@@ -59,6 +58,8 @@ def cmake_args(self):
         # See https://github.com/symengine/symengine/blob/master/README.md
         # for build options
         options.extend([
+            '-DCMAKE_BUILD_TYPE:STRING=%s' % (
```
Comment (davydden) [inline var/spack/repos/builtin/packages/symengine/package.py]: this is handled by `CMakePackage`.

File: var/spack/repos/builtin/packages/symengine/package.py (hunk 2 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/symengine/package.py]: > build_type is a standard variant for CmakePackage

Ok, I saw the `debug` variant in Trilinos and thought that must be right. I guess its just a legacy option then. I'll revert this change.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 1 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: It looks like one does need this. Without it, I get the following error (note concretising without `SymEngine`):
```
$ ./spack spec dealii@9.1~petsc~slepc+trilinos~symengine
Input spec
--------------------------------
dealii@9.1~petsc~slepc~symengine+trilinos

Concretized
--------------------------------
==> Error: multiple values are not allowed for variant "build_type"
```

I understand that this specified that if one is building with SymEngine and Trilinos enabled, then SymEngine must be built with the specified settings. With just `depends_on("symengine~debug", when="^trilinos~debug")`, I believe it is stating a dependency of SymEngine if building with Trilinos enabled, whether or not SymEngine is enabled.

File: var/spack/repos/builtin/packages/symengine/package.py (hunk 3 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/symengine/package.py]: Thanks :-)

**Comment by jppelteret** [comment]
Thanks for informing me of what I did wrong :-D

**Pull request merged**